### ISSUE TITLE
Immediate mode API (with functional flavor) for optimizers

### ISSUE THREAD

**Issue by narfanar** [pytorch/pytorch#38471]
## 🚀 Feature
_Not really formally designed. Just a seed of an idea._

Example methods:
```py
optim.apply_grad(param, **update_args)
```
Throws if `param` was never under `optim`
```py
optim.apply_grad(param, grad, **update_args)
```
Very vaguely feels like there might be something to salvage here. It feels like it would be quite awkward for stateful optimizers (Adam, etc.)
```py
ggraph = optim.new_gradient_graph()
ggraph = loss.backward(ggraph)
params, grads = select_params(ggraph)
modify_grad(grads)
optim.apply_grad(params, grads, **update_args)
```
And so on.

`update_args` would be per-optimizer usable variables. E.g. learning rate.

## Motivation

The current API is awkward to work around for extending gradient calculations or custom update regimes.

This would make customizing LR, gradient clipping, selective freezing, etc., doable through an explicit loop. Like decoupling evaluation, grad calculation, and optimizer step did for the overall training loop.

## Pitch

Let's discuss this. I don't have enough of the full picture to push it. Just felt it would be interesting.

## Alternatives

Same as today. Gradients are hidden in an associated graph, `requires_grad` loops, only in-place mutation for grad changes, etc.

## Additional context

N/A

**Comment by munael** [comment]
Relevant:

- https://github.com/facebookresearch/functorch/issues/23
- https://github.com/pytorch/pytorch/blob/master/torch/optim/_functional.py?rgh-link-date=2021-05-11T13%3A13%3A39Z
